vm/starnix/starnix.go

```diff
@@ -0,0 +1,395 @@
+	diagnose chan bool
+}
+
+func ctor(env *vmimpl.Env) (vmimpl.Pool, error) {
+	cfg := &Config{}
+	if err := config.LoadData(env.Config, cfg); err != nil {
+		return nil, fmt.Errorf("failed to parse starnix vm config: %v", err)
+	}
+	if cfg.Count < 1 || cfg.Count > 128 {
+		return nil, fmt.Errorf("invalid config param count: %v, want [1, 128]", cfg.Count)
+	}
+	if !osutil.IsDir(cfg.Fuchsia) {
+		return nil, fmt.Errorf("%q is not a valid directory", cfg.Fuchsia)
+	}
+
+	pool := &Pool{
+		count: cfg.Count,
+		env:   env,
+		cfg:   cfg,
+	}
+	return pool, nil
+}
+
+func (pool *Pool) Count() int {
+	return pool.count
+}
+
+func (pool *Pool) Create(workdir string, index int) (vmimpl.Instance, error) {
+	for i := 0; ; i++ {
+		inst, err := pool.ctor(workdir, index)
+		if err == nil {
+			return inst, nil
+		}
+		return nil, err
+	}
+}
+
+func (pool *Pool) ctor(workdir string, index int) (vmimpl.Instance, error) {
+	inst := &instance{
+		fuchsia:  pool.cfg.Fuchsia,
+		name:     fmt.Sprintf("VM-%v", index),
+		index:    index,
+		cfg:      pool.cfg,
+		debug:    pool.env.Debug,
+		timeouts: pool.env.Timeouts,
+		workdir:  workdir,
+		executor: filepath.Join(workdir, "adb_executor.sh"),
+	}
+	closeInst := inst
+	defer func() {
+		if closeInst != nil {
+			closeInst.Close()
+		}
+	}()
+
+	var err error
+	inst.rpipe, inst.wpipe, err = osutil.LongPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := inst.boot(); err != nil {
+		return nil, err
+	}
+
+	closeInst = nil
+	return inst, nil
+}
+func (inst *instance) adbPids() ([]string, error) {
+	cmd := osutil.Command("lsof", "-ti", fmt.Sprintf(":%d", inst.port))
+	rawOutput, err := cmd.Output()
+	if err != nil {
+		return nil, err
+	}
+	pidsString := strings.Trim(string(rawOutput), "\n ")
+	return strings.Split(pidsString, "\n"), nil
+}
+
+func (inst *instance) terminateFuchsiaVm() {
+	if inst.debug {
+		log.Logf(0, "Terminating Fuchsia VM: %v", inst.name)
+	}
+	inst.runCommand("ffx", "emu", "stop", inst.name)
+}
+
+func (inst *instance) Close() {
+	if inst.starnix != nil {
+		inst.terminateFuchsiaVm()
+	}
+	if inst.merger != nil {
+		inst.merger.Wait()
+	}
+	if inst.rpipe != nil {
+		inst.rpipe.Close()
+	}
+	if inst.wpipe != nil {
+		inst.wpipe.Close()
+	}
+}
+
+func (inst *instance) startFuchsiaVm() error {
+	cmd, err := inst.runCommand("ffx", "emu", "start", "--headless", "--name", inst.name)
+	if err != nil {
+		return err
+	}
+	inst.starnix = cmd
+	return nil
+}
+
+func (inst *instance) startFuchsiaLogs() error {
+	cmd := osutil.Command("ffx", "--target", inst.name, "log")
+	cmd.Dir = inst.fuchsia
+	cmd.Stdout = inst.wpipe
+	cmd.Stderr = inst.wpipe
+	inst.merger.Add("fuchsia", inst.rpipe)
+	return cmd.Start()
+}
+
+func (inst *instance) startAdbServerAndConnection(timeout time.Duration) error {
+	cmd := osutil.Command("ffx", "--target", inst.name, "starnix", "adb", "-p", fmt.Sprintf("%d", inst.port))
+	cmd.Dir = inst.fuchsia
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	inst.adb = cmd
```

Comment:
Similarly, I think this subprocess probably needs to be cleaned up on error/exit, right?

Reply:
Just checking, if `err != nil` there is no need for cleanup right?

Reply:
If Start failed, then no need to cleanup.

Reply:
Indeed, I was just referring to cleanup in instance.Close, which it looks like you've addressed.